File: 2022/day3.lua
input=io.open"day3.txt":read"*a"
n=0
t={}


for l in input:gmatch"%g+" do
  local a, b = l:sub(1,#l/2),l:sub(#l/2+1)
  local m = {}
  for i = 1, #a do
    m[a:byte(i)] = true
  end
  local m2 = {}
  for i = 1, #b do
    m2[b:byte(i)] = true
  end
  for k in pairs(m) do
    if not m2[k] then m[k] = nil end
  end
  for k, v in pairs(m) do s=k - string.byte("a") + 1 
    if s < 0 then s = s + 58 end
    n=n+s
  end
end
print(n)
n=0

for a, b, c in input:gmatch"(%g+)\n(%g+)\n(%g+)" do
  local m = {}
  for i = 1, #a do
    m[a:byte(i)] = true
  end
  local m2 = {}
  for i = 1, #b do
    m2[b:byte(i)] = true
  end
  local m3 = {}
  for i = 1, #c do
    m3[c:byte(i)] = true
  end
  for k in pairs(m) do
    if not m2[k] or not m3[k] then m[k] = nil end
  end
  for k, v in pairs(m) do s=k - string.byte("a") + 1 
    if s < 0 then s = s + 58 end
    n=n+s
  end
end

print(n)

File: 2022/day3.rb
input = File.read("day3.txt")
p input.split("\n").map{|x|
  a, b = x.chars.each_slice(x.size/2).to_a
  a & b
}.map{|x| x[0].ord - ('a'.ord - 1) }
   .map{|x| x < 0 ? x + 58 : x }.sum
a = input.lines
p a.each_slice(3)
   .map{|x| x.map{|f| f.chomp.chars }}
   .map{|r| r[0] & r[1] & r[2] }.flatten
   .map{|x| x[0].ord - ('a'.ord - 1) }
   .map{|x| x < 0 ? x + 58 : x }.sum
